File: php/src/ErrorCode.php
<?php

declare(strict_types=1);

namespace Oojs;

/**
 * String constants for all validation error codes.
 */
final class ErrorCode
{
    // Discriminator errors
    public const MISSING_DISCRIMINATOR      = 'MISSING_DISCRIMINATOR';
    public const INVALID_DISCRIMINATOR_TYPE = 'INVALID_DISCRIMINATOR_TYPE';
    public const UNKNOWN_TYPE               = 'UNKNOWN_TYPE';
    public const ABSTRACT_TYPE              = 'ABSTRACT_TYPE';
    public const TYPE_MISMATCH              = 'TYPE_MISMATCH';

    // Required property errors
    public const MISSING_REQUIRED           = 'MISSING_REQUIRED';

    // Additional property errors
    public const ADDITIONAL_PROPERTY        = 'ADDITIONAL_PROPERTY';

    // String constraint errors
    public const STRING_TOO_SHORT           = 'STRING_TOO_SHORT';
    public const STRING_TOO_LONG            = 'STRING_TOO_LONG';
    public const PATTERN_MISMATCH           = 'PATTERN_MISMATCH';
    public const ENUM_MISMATCH              = 'ENUM_MISMATCH';

    // Numeric constraint errors
    public const BELOW_MINIMUM              = 'BELOW_MINIMUM';
    public const ABOVE_MAXIMUM              = 'ABOVE_MAXIMUM';
    public const BELOW_EXCLUSIVE_MINIMUM    = 'BELOW_EXCLUSIVE_MINIMUM';
    public const ABOVE_EXCLUSIVE_MAXIMUM    = 'ABOVE_EXCLUSIVE_MAXIMUM';
    public const NOT_MULTIPLE_OF            = 'NOT_MULTIPLE_OF';
    public const NOT_INTEGER                = 'NOT_INTEGER';

    // Array constraint errors
    public const ARRAY_TOO_SHORT            = 'ARRAY_TOO_SHORT';
    public const ARRAY_TOO_LONG             = 'ARRAY_TOO_LONG';
    public const ARRAY_DUPLICATE_ITEMS      = 'ARRAY_DUPLICATE_ITEMS';

    // Graph document errors (§8.12)
    public const UNRESOLVED_REFERENCE       = 'UNRESOLVED_REFERENCE';

    private function __construct() {}
}

File: java/src/main/java/org/oojs/ErrorCode.java
package org.oojs;

/** String constants for all validation error codes. */
public final class ErrorCode {
    public static final String MISSING_DISCRIMINATOR      = "MISSING_DISCRIMINATOR";
    public static final String INVALID_DISCRIMINATOR_TYPE = "INVALID_DISCRIMINATOR_TYPE";
    public static final String UNKNOWN_TYPE               = "UNKNOWN_TYPE";
    public static final String ABSTRACT_TYPE              = "ABSTRACT_TYPE";
    public static final String TYPE_MISMATCH              = "TYPE_MISMATCH";
    public static final String MISSING_REQUIRED           = "MISSING_REQUIRED";
    public static final String ADDITIONAL_PROPERTY        = "ADDITIONAL_PROPERTY";
    public static final String STRING_TOO_SHORT           = "STRING_TOO_SHORT";
    public static final String STRING_TOO_LONG            = "STRING_TOO_LONG";
    public static final String PATTERN_MISMATCH           = "PATTERN_MISMATCH";
    public static final String ENUM_MISMATCH              = "ENUM_MISMATCH";
    public static final String BELOW_MINIMUM              = "BELOW_MINIMUM";
    public static final String ABOVE_MAXIMUM              = "ABOVE_MAXIMUM";
    public static final String BELOW_EXCLUSIVE_MINIMUM    = "BELOW_EXCLUSIVE_MINIMUM";
    public static final String ABOVE_EXCLUSIVE_MAXIMUM    = "ABOVE_EXCLUSIVE_MAXIMUM";
    public static final String NOT_MULTIPLE_OF            = "NOT_MULTIPLE_OF";
    public static final String NOT_INTEGER                = "NOT_INTEGER";
    public static final String ARRAY_TOO_SHORT            = "ARRAY_TOO_SHORT";
    public static final String ARRAY_TOO_LONG             = "ARRAY_TOO_LONG";
    public static final String ARRAY_DUPLICATE_ITEMS      = "ARRAY_DUPLICATE_ITEMS";

    private ErrorCode() {}
}
